File: ws-java/src/main/java/se/cygni/wrk/Handler.java
package se.cygni.wrk;

import rx.Observable;
import rx.Observer;
import rx.observables.ConnectableObservable;

import java.net.URI;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class Handler {

    private final DuckDuckGoClient duckDuckGoClient;

    public Handler(DuckDuckGoClient duckDuckGoClient) {
        this.duckDuckGoClient = duckDuckGoClient;
    }

    /**
     * Notifies the handler that a new connection was opened.
     * <p>
     * You are provided with a number of input {@link Observable}s as
     * well as a number of output {@link Observer}s.
     * The different {@link Observable}s represent different <b>input</b> UI controls and will emit events as these change state.
     * The different {@link Observer}s represent <b>output</b> controls. Pushing to these will change output controls in the UI.
     * <p>
     * Your task in this handler is to implement the business logic for the UI by wiring the input
     * {@link Observable}s through a Rx pipeline to to the output {@link Observer}s.
     * <p>
     * The wiring which you perform in this method will have effect throughout the lifetime of the
     * connection. The {@link Server} keeps track of the {@link Observable}s and {@link Observer}s
     * for any given connection, and since they will become wired together through your pipeline,
     * the pipeline will persist.
     * @param goClicks emits an empty string whenever the "Go" button in the GUI is clicked.
     * @param queryInputs emits query phrases from the search field. Will emit the complete query phrase whenever
     *                    it is changed in the GUI.
     * @param instantSearchChanges emits a boolean representing the checkbox "instant search" state whenever it changes.
     * @param enterPresses emits an empty string whenever the enter key is pressed in the search field.
     * @param links pushing a list of URL strings to this observer will replace the result list with the given links.
     * @param status pushing a string to this observer will update the "backend status" field.
     */
    public void onConnectionOpen(
            Observable<String> goClicks,
            Observable<String> queryInputs,
            Observable<Boolean> instantSearchChanges,
            Observable<String> enterPresses,
            Observer<List<URI>> links,
            Observer<String> status) {
        Observable<String> textOnGoClick = queryInputs.sample(goClicks);
        final Observable<String> textOnEnterPress = queryInputs.sample(enterPresses);
        final Observable<String> textOnTypeWhenInstantEnabled = Observable.combineLatest(queryInputs,
                instantSearchChanges, InstantType::new).filter(ie -> ie.instantEnabled)
                .map(ie -> ie.text);
        textOnTypeWhenInstantEnabled.map(o -> "listening").subscribe(status);
        Observable<String> debouncedTextOnTypeWhenInstantEnabled = textOnTypeWhenInstantEnabled.debounce(1, TimeUnit.SECONDS);
        Observable<String> textOnAction = textOnGoClick.mergeWith(textOnEnterPress).mergeWith(debouncedTextOnTypeWhenInstantEnabled);
        textOnAction.map(o -> "searching").subscribe(status);
        textOnAction.subscribe(s -> {
            System.out.println("About to search");
        });
        final Observable<List<URI>> requests = textOnAction.flatMap(duckDuckGoClient::searchRelated);
        //Don't want two subscribers on requests since it triggers double requests being issued
        final ConnectableObservable<List<URI>> doneRequests = requests.publish();
        doneRequests.zipWith(textOnAction, (o, text) -> "search for '" + text + "' done").subscribe(status);
        doneRequests.subscribe(links);
        doneRequests.connect();
        status.onNext("ready");
    }

    private class InstantType {
        private final String text;
        private final boolean instantEnabled;

        public InstantType(String text, boolean instantEnabled) {
            this.text = text;
            this.instantEnabled = instantEnabled;
        }
    }
}

File: ws-dotnet/src/Cygni.Competence.Reactive/Handler.cs
using System;
using System.Reactive.Linq;
using System.Reactive.Subjects;

namespace Cygni.Competence.Reactive
{
    public class Handler
    {
        private DuckDuckGoClient duckClient;

        public Handler(DuckDuckGoClient duckClient)
        {
            this.duckClient = duckClient;
        }


        /// <summary>
        /// 
        /// </summary>
        /// <param name="goClicks">emits an empty string whenever the "Go" button in the GUI is clicked.</param>
        /// <param name="queryInputs">emits query phrases from the search field. Will emit the complete query phrase whenever it is changed in the GUI</param>
        /// <param name="instantSearchChanges">emits a boolean representing the checkbox "instant search" state whenever it changes</param>
        /// <param name="enterPresses">emits an empty string whenever the enter key is pressed in the search field</param>
        /// <param name="links">pushing an array of URL strings to this observer will replace the result list with the given links</param>
        /// <param name="status">pushing a string to this observer will update the "backend status" field</param>
        public void OnConnectionOpen(
            IObservable<string> goClicks,
            IObservable<string> queryInputs,
            IObservable<bool> instantSearchChanges,
            IObservable<string> enterPresses,
            Subject<string[]> links,
            Subject<string> status)
        {
            var queryInputsWhileInstantSearchActive =
                   queryInputs.CombineLatest(instantSearchChanges, (s, instantSearch) => new { InstantSearch = instantSearch, Text = s }) //Combine any input from instant search and input
                       .Where(a => a.InstantSearch) //Where inputsearch was active
                       .Throttle(TimeSpan.FromMilliseconds(500)).Select(a => a.Text); //and ignore anything where a new keypress comes withing 500ms

            var doSearch = queryInputs.Sample(goClicks.Merge(enterPresses).Merge(queryInputsWhileInstantSearchActive)); 

            doSearch.Subscribe(
                async s =>
                {
                    if (s != String.Empty)
                    {
                        try
                        {
                            status.OnNext("Searching for " + s);
                            var results = await new DuckDuckGoClient().Search(String.Join("", s));
                            links.OnNext(results);
                            status.OnNext("Done searching for " + s);
                        }
                        catch (Exception ex)
                        {
                            status.OnNext("Error searching: " + ex.Message);
                        }
                    }
                }
            );
        }
    }
}